### PR TITLE
Add Deep Lake as an optional vector store

### PR DESCRIPTION
**Pull request by mcleonard**
This adds an extension for using Deep Lake as the vector store.

### CODE REVIEW

File: extensions/deeplake.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+import deeplake
+from deeplake.constants import MB
+import numpy as np
+import openai
+
+
+class DeepLakeStorage:
+    def __init__(self, dataset: str, token: str | None, llm_model, llm_model_path):
+        """Initialize DeepLake Storage Memory Provider"""
+        self.dataset = deeplake.empty(dataset, token=token)
```
Comment (adolkhan) [inline extensions/deeplake.py:12]: `deeplake.empty` is used either to create a new dataset or overwrite an existing dataset. To overwrite you will need to use `overwrite=True`.


But what if dataset already exists? it is better to add a check:

```suggestion
        if overwrite=True or not deeplake.exists(dataset, token=token, **creds):
             self.dataset = deeplake.empty(dataset, token=token, overwrite=True, **creds)
             
             self.dataset.create_tensor(
                 "task",
                 htype="text",
                 create_id_tensor=False,
                 create_sample_info_tensor=False,
                 create_shape_tensor=False,
                 chunk_compression="lz4",
             )

             self.dataset.create_tensor(
                 "result",
                 htype="text",
                 create_id_tensor=False,
                 create_sample_info_tensor=False,
                 create_shape_tensor=False,
                 chunk_compression="lz4",
             )

             self.dataset.create_tensor(
                 "embedding",
                 htype="generic",
                 dtype=np.float64,
                 create_id_tensor=False,
                 create_sample_info_tensor=False,
                 max_chunk_size=64 * MB,
                 create_shape_tensor=True,
             )
        else:
            self.dataset = deeplake.load(dataset, token=token, **creds)
```

File: extensions/deeplake.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        self.dataset.create_tensor(
+            "task",
```
Comment (adolkhan) [inline extensions/deeplake.py:14]: Also don't we have any kind of metadata associated with the dataset, like the date when the file was created and so on? If BabyAgi doesn't support it, it's fine we can skip it, but if we do want to add metadata, then we can create a separate tensor:

```
        self.dataset.create_tensor(
             "metadata",
             htype="json",
             create_id_tensor=False,
             create_sample_info_tensor=False,
             create_shape_tensor=False,
             chunk_compression="lz4",
         )
```

File: extensions/requirements.txt [hunk 3]
```diff
@@ -1,4 +1,5 @@
 ray==2.3.1
 pinecone-client==2.2.1
 llama-cpp-python>=0.1.35
-weaviate-client>=3.16.1
\ No newline at end of file
+weaviate-client>=3.16.1
+deeplake==3.4.1
```
Comment (adolkhan) [inline extensions/requirements.txt:5]: We will be adding bench of new features, better not to fix the version
```suggestion
deeplake>=3.4.1
```

**Review by adolkhan — CHANGES_REQUESTED** [review]

**Comment by adolkhan** [comment]
Broader comments, we need to add tests and a very simple use case on how to use babyAGI with DeepLake (also maybe part of the tests). It can even be a random data, just to make sure that it works as intended.

**Comment by mikayelh** [comment]
@yoheinakajima, i wonder if this can be reviewed on your end as well - BabyAgi would benefit from another OSS vector store! (we're also working on a use case that's based on this integration).

**Pull request closed without merging**